### PR TITLE
8313874 - JNI NewWeakGlobalRef throws exception for null arg

### PR DESCRIPTION
**Pull request by olivergillespie**
According to the JNI spec for NewWeakGlobalRef:

> Returns NULL if obj refers to null, or if obj was a weak global reference, or if the VM runs out of memory. If the VM runs out of memory, an OutOfMemoryError will be thrown.

The current behaviour is that `NewWeakGlobalRef(nullptr)` throws an OutOfMemoryError, which it should not.

(Thanks @dholmes-ora for confirming the issue and suggesting a fix.)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313874](https://bugs.openjdk.org/browse/JDK-8313874): JNI NewWeakGlobalRef throws exception for null arg (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15188/head:pull/15188` \
`$ git checkout pull/15188`

Update a local copy of the PR: \
`$ git checkout pull/15188` \
`$ git pull https://git.openjdk.org/jdk.git pull/15188/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15188`

View PR using the GUI difftool: \
`$ git pr show -t 15188`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15188.diff">https://git.openjdk.org/jdk/pull/15188.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15188#issuecomment-1669381026)

### CODE REVIEW

**Comment by stefank** [comment]
> or if obj was a weak global reference

There's no check to see if `obj` is a weak global reference in this function. Is that also an oversight or is the spec just weirdly phrased?

File: src/hotspot/share/prims/jni.cpp [hunk 1]
```diff
@@ -2876,7 +2876,7 @@ JNI_ENTRY(jweak, jni_NewWeakGlobalRef(JNIEnv *env, jobject ref))
   HOTSPOT_JNI_NEWWEAKGLOBALREF_ENTRY(env, ref);
   Handle ref_handle(thread, JNIHandles::resolve(ref));
   jweak ret = JNIHandles::make_weak_global(ref_handle, AllocFailStrategy::RETURN_NULL);
-  if (ret == nullptr) {
+  if (ret == nullptr && ref_handle != nullptr) {
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jni.cpp:2879]: That isn't how you check a Handle for null content - you need `!ref_handle.is_null()`.

File: test/hotspot/jtreg/runtime/jni/NullJNIWeak/NullJNIWeak.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/jni/NullJNIWeak/NullJNIWeak.java:1]: Is there not an existing test we can simply add the null case to?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: src/hotspot/share/prims/jni.cpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/prims/jni.cpp:2879]: even better would be `ref_handle.not_null()`.  Though I think the comparison does work.

File: test/hotspot/jtreg/runtime/jni/NullJNIWeak/NullJNIWeak.java
Comment (kimbarrett) [reply test/hotspot/jtreg/runtime/jni/NullJNIWeak/NullJNIWeak.java:1]: Agree that adding this to some existing test might be preferable.  Maybe another subtest (see main) in
runtime/jni/ReturnJNIWeak?

**Review by kimbarrett — CHANGES_REQUESTED** [review]

**Comment by stefank** [comment]
I see that the spec further down says:
`obj was a weak global reference and has already been garbage collected`

So, it is just the first `obj was a weak global reference` occurence that is ill-worded.

File: src/hotspot/share/prims/jni.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jni.cpp]: Yes it seems that some implicit Handle conversions were removed at some point which now makes direct comparison to `nullptr` allowable - somewhat obsolescing the `is_null()/not_null()` helper methods.

**Review by shipilev — APPROVED** [review]
Looks good to me.

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks.

I will file a follow up RFE to look at the spec issue @stefank  pointed out.

**Pull request closed without merging**